src/Dialogs/Device/FLARM/RangeConfigWidget.cpp

```diff
@@ -0,0 +1,131 @@
+  "ADSBRANGE",
+  "ADSBVRANGE",
+  NULL
+};
+
+static unsigned
+GetUnsignedValue(const FlarmDevice &device, const char *name,
+                 unsigned default_value)
+{
+  if (const auto x = device.GetSetting(name)) {
+    char *endptr;
+    unsigned long y = strtoul(x->c_str(), &endptr, 10);
+    if (endptr > x->c_str() && *endptr == 0)
+      return (unsigned)y;
+  }
+
+  return default_value;
+}
+
+void
+FLARMRangeConfigWidget::Prepare([[maybe_unused]] ContainerWindow &parent,
+                           [[maybe_unused]] const PixelRect &rc) noexcept
+{
+  PopupOperationEnvironment env; 
+  device.RequestAllSettings(flarm_setting_names, env);
+  if (hardware.isPowerFlarm())
+    device.RequestAllSettings(pf_setting_names, env);
+  if (hardware.hasADSB())
+    device.RequestAllSettings(adsb_setting_names, env);
+
+  range = GetUnsignedValue(device, "RANGE", 3000);
+  vrange = GetUnsignedValue(device, "VRANGE", 500);
+  pcas_range = GetUnsignedValue(device, "PCASRANGE", 7408);
+  pcas_vrange = GetUnsignedValue(device, "PCASVRANGE", 610);
+  adsb_range = GetUnsignedValue(device, "ADSBRANGE", 65535);
+  adsb_vrange = GetUnsignedValue(device, "ADSBVRANGE", 65535);
+
+  AddInteger(_("Range"), NULL, _T("%d m"), _T("%d"), 2000, 25500, 250, range);
+
+  if (hardware.isPowerFlarm()) {
+    AddInteger(_("Vertical range"), NULL, _T("%d m"), _T("%d"), 100, 2000, 100, vrange);
+  }
+  if (hardware.hasADSB()) {
+    AddInteger(_("PCAS range"), NULL, _T("%d m"), _T("%d"), 500, 9260, 500, pcas_range);
+    AddInteger(_("PCAS vertical range"), NULL, _T("%d m"), _T("%d"), 250, 65535, 250, pcas_vrange);
+    AddInteger(_("ADSB range"), NULL, _T("%d m"), _T("%d"), 500, 65535, 500, adsb_range);
+    AddInteger(_("ADSB vertical range"), NULL, _T("%d m"), _T("%d"), 250, 65535, 250, adsb_vrange);
+  }
+}
+
+bool
+FLARMRangeConfigWidget::Save(bool &_changed) noexcept
+try {
+  PopupOperationEnvironment env;
+  bool changed = false;
+  NarrowString<32> buffer;
+
+  if (SaveValueInteger(Range, range)) {
+    buffer.UnsafeFormat("%u", range);
+    device.SendSetting("RANGE", buffer, env);
+    changed = true;
+  }
+
+  if (SaveValueInteger(VRange, vrange)) {
+    buffer.UnsafeFormat("%u", vrange);
+    device.SendSetting("VRANGE", buffer, env);
+    changed = true;
+  }
+
+  if (SaveValueInteger(PCASRange, pcas_range)) {
+    buffer.UnsafeFormat("%u", pcas_range);
+    device.SendSetting("PCASRANGE", buffer, env);
+    changed = true;
+  }
+
+  if (SaveValueInteger(PCASVRange, pcas_vrange)) {
+    buffer.UnsafeFormat("%u", pcas_vrange);
+    device.SendSetting("PCASVRANGE", buffer, env);
+    changed = true;
+  }
+
+  if (SaveValueInteger(ADSBRange, adsb_range)) {
+    buffer.UnsafeFormat("%u", adsb_range);
+    device.SendSetting("ADSBRANGE", buffer, env);
+    changed = true;
+  }
+
+  if (SaveValueInteger(ADSBVrange, adsb_vrange)) {
+    buffer.UnsafeFormat("%u", adsb_vrange);
+    device.SendSetting("ADSBVRANGE", buffer, env);
+    changed = true;
+  }
```

Comment:
These need to be gated by isPowerFlarm() and hasADSB(); Otherwise xcsoar will trigger the mentioned assertion when saving the range on a classic flarm.

Reply:
done